nav2_ceres_costaware_smoother/src/ceres_costaware_smoother.cpp

```diff
@@ -0,0 +1,156 @@
+#include "nav2_costmap_2d/costmap_filters/filter_values.hpp"
+
+#include "pluginlib/class_loader.hpp"
+#include "pluginlib/class_list_macros.hpp"
+
+using nav2_util::declare_parameter_if_not_declared;
+using nav2_util::geometry_utils::euclidean_distance;
+using namespace nav2_costmap_2d;  // NOLINT
+
+namespace nav2_ceres_costaware_smoother
+{
+
+void CeresCostawareSmoother::configure(
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  std::string name, const std::shared_ptr<tf2_ros::Buffer> & tf,
+  const std::shared_ptr<nav2_costmap_2d::CostmapSubscriber> & costmap_sub,
+  const std::shared_ptr<nav2_costmap_2d::FootprintSubscriber> &)
+{
+  auto node = parent.lock();
+  if (!node) {
+    throw std::runtime_error("Unable to lock node!");
+  }
+
+  costmap_sub_ = costmap_sub;
+  tf_ = tf;
+  plugin_name_ = name;
+  logger_ = node->get_logger();
+
+  _smoother = std::make_unique<nav2_ceres_costaware_smoother::Smoother>();
+  _optimizer_params.get(node.get(), name);
+  _smoother_params.get(node.get(), name);
+  _smoother->initialize(_optimizer_params);
+}
+
+void CeresCostawareSmoother::cleanup()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Cleaning up smoother: %s of type"
+    " nav2_ceres_costaware_smoother::CeresCostawareSmoother",
+    plugin_name_.c_str());
+}
+
+void CeresCostawareSmoother::activate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Activating smoother: %s of type "
+    "nav2_ceres_costaware_smoother::CeresCostawareSmoother",
+    plugin_name_.c_str());
+}
+
+void CeresCostawareSmoother::deactivate()
+{
+  RCLCPP_INFO(
+    logger_,
+    "Deactivating smoother: %s of type "
+    "nav2_ceres_costaware_smoother::CeresCostawareSmoother",
+    plugin_name_.c_str());
+}
+
+bool CeresCostawareSmoother::smooth(nav_msgs::msg::Path & path, const rclcpp::Duration & max_time)
+{
+  if (path.poses.size() < 2) {
+    return true;
+  }
+
+  auto costmap = costmap_sub_->getCostmap();
+  std::unique_lock<nav2_costmap_2d::Costmap2D::mutex_t> lock(*(costmap->getMutex()));
+
+  // populate smoother input with (x, y, forward/reverse dir)
+  std::vector<Eigen::Vector3d> path_world;
+  path_world.reserve(path.poses.size());
+  Eigen::Vector2d start_dir;
+  Eigen::Vector2d end_dir;
+  for (size_t i = 0; i < path.poses.size(); i++) {
+    auto & pose = path.poses[i].pose;
+    tf2::Quaternion q;
+    tf2::fromMsg(pose.orientation, q);
+    double angle = q.getAngle() * sign(q.getZ());
+    Eigen::Vector2d orientation(cos(angle), sin(angle));
+    if (i == path.poses.size() - 1) {
+      path_world.emplace_back(pose.position.x, pose.position.y, path_world.back()[2]);
+      end_dir = orientation;
+    } else {
+      auto & pos_next = path.poses[i + 1].pose.position;
+      Eigen::Vector2d mvmt(pos_next.x - pose.position.x, pos_next.y - pose.position.y);
+      bool reversing = _smoother_params.reversing_enabled && orientation.dot(mvmt) < 0;
+      path_world.emplace_back(pose.position.x, pose.position.y, reversing ? -1 : 1);
```

Comment:
The reversing logic is a little bit convoluted, it would be simpler to do something like what we do here: https://github.com/ros-planning/navigation2/blob/fee433ff76b1fe2997043bec16b81568fbead49e/nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp#L646-L673 to use the dot product of the robot poses to find cusp point of directional changes.

Reply:
It is a little bit more convoluted but much more robust. Frankly, I know this piece of code and have had bad experiences with it. (Haven't raised an issue though yet, I preferred returning back to our own controller plugin, since there was more than this to change in the controller to suit our needs).

If you think about it, a path cusp is a point where a forward movement direction changes to backward (that's what our implementation checks for). Although usually, but not necessarily (e.g. when sampling is not dense enough), it shows a >90° break in the path (that's what the RPPC implementation checks for). I understand that our attitude requires orientations to be assigned in the input path (or `reversing_enabled` parameter to be turned off) but that's what usually planners which can produce backward motions do provide, so eliminating the requirement is not worth stepping back with the robustness.

Reply:
OK - I think more than the actual algorithm used its the lack of standard tools used to implement it. More than I have an issue with what this code does, my concern is that the meaning / semantics of what you're trying to accomplish are lost to the passive reader. 

```
    tf2::Quaternion q;
    tf2::fromMsg(pose.orientation, q);
    double angle = q.getAngle() * sign(q.getZ());
    Eigen::Vector2d orientation(cos(angle), sin(angle));
    ...
    orientation.dot(mvmt)
    reversing ? -1 : 1
```
Can this be simplified / use more standard tools (or add comments?) so that the intention of what's happening is clear?

Reply:
I think you are looking at an outdated commit... the tf2::getYaw is currently used there (reason I used getAngle and getZ was that this was one of the first things I did in ROS2 so I didn't know about all the misc utility methods, so I just sticked to class member methods, which were all at one place, considering that planner output paths are 2D...)

Adding clarifying comments to other mentioned operations...

Reply:
You are correct, yes this is much more clear now, thank you.